File: c/dnd-character/dnd_character.c
#include "dnd_character.h"

#include <stdlib.h>
#include <time.h>

#define MAX_ROLL 6
#define NUM_ROLLS 4

int modifier(int score) {
    int modifier = (score - 10) / 2;
    return modifier - (score <= 10 && score % 2 != 0);
}

int ability(void) {
    srand(time(NULL));

    int min = MAX_ROLL + 1;
    int sum = 0;
    for (int i = 0; i < NUM_ROLLS; i++) {
        int roll = rand() % MAX_ROLL + 1;
        sum += roll;
        if (roll < min) {
            min = roll;
        }
    }
    return sum - min;
}

dnd_character_t make_dnd_character(void) {
    dnd_character_t new_char;

    new_char.strength = ability();
    new_char.dexterity = ability();
    new_char.constitution = ability();
    new_char.intelligence = ability();
    new_char.wisdom = ability();
    new_char.charisma = ability();
    new_char.hitpoints = modifier(new_char.constitution) + 10;

    return new_char;
}

File: zig/dnd-character/dnd_character.zig
const rand = @import("std").Random;

pub fn modifier(score: i8) i8 {
    return @divFloor(score - 10, 2);
}

pub fn ability() i8 {
    var prng = rand.DefaultPrng.init(0);
    const r = prng.random();

    var min: i8 = 7;
    var sum: i8 = 0;
    for (0..4) |_| {
        const roll = r.intRangeAtMost(i8, 1, 6);
        sum += roll;
        if (roll < min) min = roll;
    }
    return sum - min;
}

pub const Character = struct {
    strength: i8 = ability(),
    dexterity: i8 = ability(),
    constitution: i8 = ability(),
    intelligence: i8 = ability(),
    wisdom: i8 = ability(),
    charisma: i8 = ability(),
    hitpoints: i8 = undefined,

    pub fn init() Character {
        var char = Character{};
        char.hitpoints = modifier(char.constitution) + 10;
        return char;
    }
};
